File: conformance/ruby/shapes.rb
# frozen_string_literal: true
# Conformance consumer: shapes sample, Ruby target.
#
# Drives the generated rich (algebraic) enum surface: Shape is a sum type of
# plain value classes (Shape::Empty, Shape::Circle, Shape::Rectangle, and
# Shape::Labeled), each with keyword construction, field readers, an integer
# `tag` reader backed by per-variant TAG constants, and structural equality.
# The free functions encode the shape into a value buffer on the way in and
# decode the returned buffer into a fresh variant instance. Also covers the
# expanded numerics (f32 fields, u8 field, bytes in, u64 out). The cdylib is
# selected via WEAVEFFI_LIBRARY. Non-zero exit on any failed assertion.

$LOAD_PATH.unshift(File.join(ENV.fetch("WV_RB"), "lib"))
require "shapes"

def expect(cond, msg)
  raise "assertion failed: #{msg}" unless cond
end

# Unit variant: no fields, tag 0.
empty = Shapes::Shape::Empty.new
expect(empty.is_a?(Shapes::Shape), "Empty is a Shape")
expect(empty.tag == Shapes::Shape::Empty::TAG, "empty tag is the TAG constant (got #{empty.tag})")
expect(empty.tag.zero?, "empty tag == 0 (got #{empty.tag})")
expect(empty == Shapes::Shape::Empty.new, "unit variants compare structurally")

# f64 payload.
circle = Shapes::Shape::Circle.new(radius: 2.5)
expect(circle.tag == Shapes::Shape::Circle::TAG, "circle tag is the TAG constant (got #{circle.tag})")
expect(circle.tag == 1, "circle tag == 1 (got #{circle.tag})")
expect((circle.radius - 2.5).abs < 1e-9, "circle radius == 2.5 (got #{circle.radius})")

# Two f32 payloads.
rect = Shapes::Shape::Rectangle.new(width: 3.0, height: 4.0)
expect(rect.tag == 2, "rectangle tag == 2 (got #{rect.tag})")
expect((rect.width - 3.0).abs < 1e-6, "rectangle width == 3.0 (got #{rect.width})")
expect((rect.height - 4.0).abs < 1e-6, "rectangle height == 4.0 (got #{rect.height})")

# string + u8 payload.
labeled = Shapes::Shape::Labeled.new(label: "hex", count: 6)
expect(labeled.tag == 3, "labeled tag == 3 (got #{labeled.tag})")
expect(labeled.label == "hex", "labeled label == hex (got #{labeled.label.inspect})")
expect(labeled.count == 6, "labeled count == 6 (got #{labeled.count})")

# Free functions: Shape in (encoded into a value buffer), string/Shape out.
expect(Shapes.describe(circle) == "circle(r=2.5)", "describe(circle) (got #{Shapes.describe(circle).inspect})")
expect(Shapes.describe(rect).include?("rect"), "describe(rect) mentions rect")
expect(Shapes.describe(labeled).include?("hex"), "describe(labeled) mentions the label")

big = Shapes.scale(circle, 4.0)
expect(big.is_a?(Shapes::Shape::Circle), "scaled variant is Circle (got #{big.class})")
expect((big.radius - 10.0).abs < 1e-9, "scaled radius == 10.0 (got #{big.radius})")
expect(big == Shapes::Shape::Circle.new(radius: 10.0), "scaled shape compares structurally")

# A C-style enum keeps its plain integer constants.
expect(Shapes::Channel::GREEN == 1, "Channel::GREEN == 1")

# Numerics: `[u8]` canonicalizes to `bytes`, so the parameter is a binary
# string; the u64 sum comes back as a plain Integer.
raw = [250, 250, 250, 250].pack("C*")
expect(Shapes.sum_bytes(raw) == 1000, "sum_bytes == 1000 (got #{Shapes.sum_bytes(raw)})")

puts "ruby/shapes: OK"

File: conformance/python/shapes.py
"""Conformance consumer: shapes sample, Python target.

Drives the generated ctypes wrapper for rich (algebraic) enums as value
types: the `Shape` base class with its nested `Tag` IntEnum and `tag`
reader, the per-variant dataclasses reachable as `Shape.Circle(...)` with
natural field names (`radius`, not `circle_radius`) and value equality,
plus the free functions that take and return `Shape` as value buffers
(module-prefix-stripped: `describe`, `scale`, `sum_bytes`). Also covers the
expanded numerics (f32 fields, u8 field, u64 return) and the plain C-style
`Channel` enum. The generated package is placed on sys.path via WV_PY; the
cdylib is selected with WEAVEFFI_LIBRARY.
"""
import os
import sys

sys.path.insert(0, os.environ["WV_PY"])

import shapes as wv  # noqa: E402


def main() -> None:
    # Unit variant: tag only.
    empty = wv.Shape.Empty()
    assert isinstance(empty, wv.Shape)
    assert empty.tag == wv.Shape.Tag.Empty

    # f64 payload.
    circle = wv.Shape.Circle(2.5)
    assert circle.tag == wv.Shape.Tag.Circle
    assert abs(circle.radius - 2.5) < 1e-9

    # Two f32 payloads.
    rect = wv.Shape.Rectangle(3.0, 4.0)
    assert rect.tag == wv.Shape.Tag.Rectangle
    assert abs(rect.width - 3.0) < 1e-6
    assert abs(rect.height - 4.0) < 1e-6

    # string + u8 payload.
    labeled = wv.Shape.Labeled("hex", 6)
    assert labeled.tag == wv.Shape.Tag.Labeled
    assert labeled.label == "hex"
    assert labeled.count == 6

    # Free functions: Shape in, string/Shape out.
    assert wv.describe(circle) == "circle(r=2.5)"

    big = wv.scale(circle, 4.0)
    assert isinstance(big, wv.Shape.Circle)
    assert big.tag == wv.Shape.Tag.Circle
    assert abs(big.radius - 10.0) < 1e-9

    # Variants are dataclasses: values decoded from the producer compare
    # equal to locally constructed ones.
    assert big == wv.Shape.Circle(10.0)
    assert wv.Shape.Labeled("hex", 6) == labeled

    # A C-style enum keeps its plain integer constants.
    assert wv.Channel.Green == 1

    # Numerics: `[u8]` canonicalizes to `bytes` in, u64 out.
    assert wv.sum_bytes(bytes([250, 250, 250, 250])) == 1000

    print("python/shapes: OK")


main()
